v1/plugin/stream_collector_proxy.go

```diff
@@ -1,119 +1,179 @@
 			fmt.Println(err.Error())
 		}
-
 	}
 }
 
-func metricSend(
-	plugin StreamCollector,
-	ch chan []Metric,
-	s rpc.StreamCollector_StreamMetricsServer) {
-	for r := range ch {
-		mts := []*rpc.Metric{}
-		for _, mt := range r {
-			metric, err := toProtoMetric(mt)
-			if err != nil {
-				fmt.Println(err.Error())
+func (p *StreamProxy) metricSend(ch chan []Metric, stream rpc.StreamCollector_StreamMetricsServer) {
+	metrics := []*rpc.Metric{}
+
+	for {
+		select {
+		case mts := <-ch:
+			if len(mts) == 0 {
+				break
 			}
-			mts = append(mts, metric)
-		}
-		reply := &rpc.CollectReply{
-			Metrics_Reply: &rpc.MetricsReply{Metrics: mts},
-		}
-		if err := s.Send(reply); err != nil {
-			fmt.Println(err.Error())
+
+			for _, mt := range mts {
+				metric, err := toProtoMetric(mt)
+				if err != nil {
+					fmt.Println(err.Error())
+					break
+				}
+				metrics = append(metrics, metric)
+
+				// send metrics if maxMetricsBuffer is reached
+				// (notice it is only possible for maxMetricsBuffer greater than 0)
+				if p.maxMetricsBuffer == int64(len(metrics)) {
+					sendReply(metrics, stream)
+					metrics = []*rpc.Metric{}
+				}
+			}
+
+			// send all available metrics immediately for maxMetricsBuffer is 0 (defaults)
+			if p.maxMetricsBuffer == 0 {
+				sendReply(metrics, stream)
+				metrics = []*rpc.Metric{}
+			}
+		case <-time.After(p.maxCollectDuration):
+			// send metrics if maxCollectDuration is reached
+			sendReply(metrics, stream)
+			metrics = []*rpc.Metric{}
+
 		}
 	}
-
 }
 
-func streamRecv(
-	plugin StreamCollector,
-	ch chan []Metric,
-	s rpc.StreamCollector_StreamMetricsServer) {
-
+func (p *StreamProxy) streamRecv(ch chan []Metric, stream rpc.StreamCollector_StreamMetricsServer) {
 	for {
-		s, err := s.Recv()
+		s, err := stream.Recv()
 		if err != nil {
 			fmt.Println(err)
 			continue
 		}
 		if s != nil {
+			if s.MaxMetricsBuffer > 0 {
+				p.setMaxMetricsBuffer(s.MaxMetricsBuffer)
+			}
+			if s.MaxCollectDuration > 0 {
+				p.setMaxCollectDuration(time.Duration(s.MaxCollectDuration))
+			}
 			if s.Metrics_Arg != nil {
 				metrics := []Metric{}
 				for _, mt := range s.Metrics_Arg.Metrics {
 					metric := fromProtoMetric(mt)
 					metrics = append(metrics, metric)
 				}
+				// send requested metrics to be collected into the stream plugin
 				ch <- metrics
 			}
 		}
 	}
 }
 
-func (c *StreamProxy) SetConfig(context.Context, *rpc.ConfigMap) (*rpc.ErrReply, error) {
-	return nil, nil
+func (p *StreamProxy) setMaxCollectDuration(d time.Duration) {
+	p.maxCollectDuration = d
 }
 
-func (c *StreamProxy) GetMetricTypes(ctx context.Context, arg *rpc.GetMetricTypesArg) (*rpc.MetricsReply, error) {
-	cfg := fromProtoConfig(arg.Config)
+func (p *StreamProxy) setMaxMetricsBuffer(i int64) {
+	p.maxMetricsBuffer = i
+}
 
-	r, err := c.plugin.GetMetricTypes(cfg)
-	if err != nil {
-		return nil, err
+func sendReply(metrics []*rpc.Metric, stream rpc.StreamCollector_StreamMetricsServer) {
```

Comment:
**To reviewers**:
Please notice that `sendReply` correspond to the original code in [v1/plugin/stream_collector_proxy.go#L63-67] (https://github.com/intelsdi-x/snap-plugin-lib-go/pull/77/files#diff-aa6fc80eb3b8c2400af503a0443c72b7L63)